File: test_libgrpp_f90/rpp.f90
!
!  libgrpp - a library for the evaluation of integrals over
!            generalized relativistic pseudopotentials.
!
!  Copyright (C) 2021-2023 Alexander Oleynichenko
!

module ecp

    use xyz
    implicit none

    integer, parameter :: ECP_MAXL = 10
    integer, parameter :: ECP_MAX_CNTRCT_LEN = 50
    integer, parameter :: ECP_MAX_OC_SHELLS = 20

    ! effective core potential data
    integer :: n_arep(N_ELEMENTS)
    integer :: n_esop(N_ELEMENTS)
    integer :: n_oc_shells(N_ELEMENTS, ECP_MAXL)

    integer :: ecp_num_prim(N_ELEMENTS, ECP_MAXL)
    real(8) :: ecp_alpha(N_ELEMENTS, ECP_MAXL, ECP_MAX_CNTRCT_LEN)
    integer :: ecp_powers(N_ELEMENTS, ECP_MAXL, ECP_MAX_CNTRCT_LEN)
    real(8) :: arep(N_ELEMENTS, ECP_MAXL, ECP_MAX_CNTRCT_LEN)
    real(8) :: esop(N_ELEMENTS, ECP_MAXL, ECP_MAX_CNTRCT_LEN)
    real(8) :: ocpot(N_ELEMENTS, ECP_MAXL, ECP_MAX_OC_SHELLS, ECP_MAX_CNTRCT_LEN)

    integer :: ocbas_num_prim(N_ELEMENTS, ECP_MAXL)
    integer :: ocbas_num_contr(N_ELEMENTS, ECP_MAXL)
    real(8) :: ocbas_alpha(N_ELEMENTS, ECP_MAXL, ECP_MAX_CNTRCT_LEN)
    real(8) :: ocbas_coeffs(N_ELEMENTS, ECP_MAXL, ECP_MAX_CNTRCT_LEN, ECP_MAX_CNTRCT_LEN)

contains

    subroutine read_rpp(basis_file, charge)

        implicit none

        ! arguments
        character(len = *), intent(in) :: basis_file
        integer, intent(in) :: charge

        ! local variables
        integer :: inp
        integer :: i
        integer :: z
        integer :: nblocks
        character :: star
        integer :: iblock
        integer :: nprim
        integer :: ncontr
        integer :: iprim
        integer :: max_oc
        integer :: n_arep_tmp
        integer :: n_oc_pot
        integer :: ioc
        integer :: irep

        n_arep(charge) = 0
        n_esop(charge) = 0
        n_oc_shells(charge, :) = 0

        inp = 10
        open(unit = inp, file = basis_file, status = 'old', err = 13)

        do while (.TRUE.)
            11 continue
            read (inp, *, end = 12, err = 11) star, z, max_oc, n_arep_tmp
            if (star == '*' .and. z == charge) then

                n_arep(z) = n_arep_tmp
                n_esop(z) = n_arep(z)

                do ioc = 1, max_oc
                    read (inp, *, err = 14) nprim, ncontr
                    ocbas_num_prim(z, ioc) = nprim
                    ocbas_num_contr(z, ioc) = ncontr
                    n_oc_shells(z, ioc) = ncontr
                    do iprim = 1, nprim
                        read (inp, *, err = 14) ocbas_alpha(z, ioc, iprim), ocbas_coeffs(z, ioc, iprim, 1:ncontr)
                    end do
                end do

                do irep = 1, n_arep(z)
                    read (inp, *, err = 14) nprim, n_oc_pot
                    ecp_num_prim(z, irep) = nprim
                    do iprim = 1, nprim
                        read (inp, *, err = 14) ecp_powers(z, irep, iprim), ecp_alpha(z, irep, iprim), &
                                arep(z, irep, iprim), esop(z, irep, iprim), &
                                (ocpot(z, irep, ioc, iprim), ioc = 1, n_oc_pot)
                    end do
                end do

                exit

            end if
        end do

        12 continue
        close(unit = inp)
        return

        13 continue
        print '(a,i3,a)', 'Input file with ECP for element ', charge, ' is not found'
        stop
        14 continue
        print *, 'Syntax error in the ECP file'
        stop

    end subroutine read_rpp

    subroutine print_ecp(charge)

        implicit none

        ! arguments
        integer, intent(in) :: charge

        character, parameter :: symbols_table(ECP_MAXL) = (/ 'S', 'P', 'D', 'F', 'G', 'H', 'I', 'K', 'L', 'M' /)
        integer :: iblock
        integer :: iprim
        integer :: irep
        integer :: ioc

        print *
        print '(a,i3,a)', ' generalized relativistic pseudopotential for element Z = ', charge, ':'
        print *, '-------------------------------------------------------------'

        print *, 'outercore basis set:'
        do irep = 1, n_arep(charge)
            if (n_oc_shells(charge, irep) == 0) then
                cycle
            end if
            print *, symbols_table(irep)
            do iprim = 1, ocbas_num_prim(charge, irep)
                print '(f20.10,a,20f16.10)', ocbas_alpha(charge, irep, iprim), '  ', &
                        ocbas_coeffs(charge, irep, iprim, 1:ocbas_num_contr(charge, irep))
            end do

        end do
        print *

        print *, 'semilocal rpp:'
        do irep = 1, n_arep(charge)
            print '(2x,a,31x,a,16x,a)', symbols_table(irep), 'arep', 'esop'
            do iprim = 1, ecp_num_prim(charge, irep)
                print '(i3,3f20.12)', ecp_powers(charge, irep, iprim), ecp_alpha(charge, irep, iprim), &
                        arep(charge, irep, iprim), esop(charge, irep, iprim)
            end do
        end do
        print *

        print *, 'outercore potentials:'
        do irep = 1, n_arep(charge)
            if (n_oc_shells(charge, irep) == 0) then
                cycle
            end if

            ! header line: LJ symbols
            if (irep == 1) then
                print '(34x,10(a,16x))', ('S1/2', ioc = 1, n_oc_shells(charge, irep))
            else
                write (*, '(35x)', advance = 'no')
                do ioc = 1, n_oc_shells(charge, irep), 2
                    write (*, '(2(a,i1,a,16x))', advance = 'no') &
                            symbols_table(irep), 2 * (irep - 1) - 1, '/2', &
                            symbols_table(irep), 2 * (irep - 1) + 1, '/2'
                end do
                print *
            end if

            do iprim = 1, ecp_num_prim(charge, irep)
                print '(i3,20f20.12)', ecp_powers(charge, irep, iprim), ecp_alpha(charge, irep, iprim), &
                        (ocpot(charge, irep, ioc, iprim), ioc = 1, n_oc_shells(charge, irep))
            end do
        end do
        print *

    end subroutine print_ecp

end module ecp

File: test_libgrpp_c/rpp.c
/*
 *  libgrpp - a library for the evaluation of integrals over
 *            generalized relativistic pseudopotentials.
 *
 *  Copyright (C) 2021-2023 Alexander Oleynichenko
 */

#include "rpp.h"

#include <stdlib.h>
#include <string.h>

int skip_line(FILE *fp);
void print_rpp_expansion(FILE *out, libgrpp_potential_t *ecp);


libgrpp_grpp_t *read_grpp(char *path, int nuc_charge)
{
    FILE *inp_file;
    char buf[256];
    int z, n_blocks, n_arep;
    const int MAX_N_EXPONENTS = 100;
    const int MAX_N_CONTRACTED = 100;
    int pow_buf[MAX_N_EXPONENTS];
    double exp_buf[MAX_N_EXPONENTS];
    double coef_buf[MAX_N_CONTRACTED][MAX_N_EXPONENTS];

    // open file with basis set
    inp_file = fopen(path, "r");
    if (inp_file == NULL) {
        printf("Cannot open ecp file '%s\n'", path);
        return NULL;
    }

    // find the entry for the required element (nuclear charge)
    int found = 0;
    while (fscanf(inp_file, "%s", buf) == 1) {
        if (strcmp(buf, "*") == 0) {

            if (fscanf(inp_file, "%d%d%d", &z, &n_blocks, &n_arep) != 3) {
                printf("Error while reading ecp file for Z = %d\n", nuc_charge);
                return NULL;
            }

            if (z == nuc_charge) {
                found = 1;
                break;
            }
        }
    }

    if (found == 0) {
        return NULL;
    }

    // create template for the effective core potential
    libgrpp_grpp_t *grpp = libgrpp_new_grpp();

    int n_oc_shells = 0;
    libgrpp_shell_t *buf_oc_shells[MAX_NUM_OC_SHELLS];
    libgrpp_potential_t *buf_oc_potentials[MAX_NUM_OC_SHELLS];

    // read outercore pseudospinor expansions
    for (int iblock = 0; iblock < n_blocks; iblock++) {
        int n_alpha;
        int n_contr;
        double origin[] = {0, 0, 0};

        fscanf(inp_file, "%d%d", &n_alpha, &n_contr);
        skip_line(inp_file);

        for (int row = 0; row < n_alpha; row++) {
            fscanf(inp_file, "%lf", &exp_buf[row]);
            for (int col = 0; col < n_contr; col++) {
                fscanf(inp_file, "%lf", &coef_buf[col][row]);
            }
        }

        // add basis functions to the set
        for (int ifun = 0; ifun < n_contr; ifun++) {
            libgrpp_shell_t *shell = libgrpp_new_shell(origin, iblock, n_alpha, coef_buf[ifun], exp_buf);
            buf_oc_shells[n_oc_shells++] = shell;
        }
    }

    // read ECP expansions
    //grpp->n_oc_shells = 0;
    n_oc_shells = 0;
    for (int iblock = 0; iblock < n_arep; iblock++) {
        int n_prim;
        int n_oc;
        int L = iblock;

        fscanf(inp_file, "%d%d", &n_prim, &n_oc);
        skip_line(inp_file);


        for (int row = 0; row < n_prim; row++) {

            // power, exponent
            fscanf(inp_file, "%d%lf", &pow_buf[row], &exp_buf[row]);

            // arep, esop
            fscanf(inp_file, "%lf%lf", &coef_buf[0][row], &coef_buf[1][row]);

            // outercore potentials
            for (int ioc = 0; ioc < n_oc; ioc++) {
                fscanf(inp_file, "%lf", &coef_buf[2 + ioc][row]);
            }

            skip_line(inp_file);
        }

        // add potentials to the GRECP structure

        libgrpp_potential_t *arep = libgrpp_new_potential(L, 0, n_prim, pow_buf, coef_buf[0], exp_buf);
        if (L == n_arep - 1) { // special case of the U_L radially-local potential
            libgrpp_grpp_set_local_potential(grpp, arep);
        }
        else {
            libgrpp_grpp_add_averaged_potential(grpp, arep);
        }

        if (L == 0) {
            // no SO term for angular momentum S => put NULL
            libgrpp_grpp_add_spin_orbit_potential(grpp, NULL);
        }
        else {
            libgrpp_potential_t *esop = libgrpp_new_potential(L, 0, n_prim, pow_buf, coef_buf[1], exp_buf);
            libgrpp_grpp_add_spin_orbit_potential(grpp, esop);
        }

        // save OC potentials to buffer
        for (int ioc = 0; ioc < n_oc; ioc++) {
            int J = abs(2 * L + ((ioc % 2 == 0) ? -1 : +1));
            libgrpp_potential_t *oc_pot = libgrpp_new_potential(L, J, n_prim, pow_buf, coef_buf[2 + ioc], exp_buf);
            buf_oc_potentials[n_oc_shells++] = oc_pot;
        }
    }

    // add pairs (OC potential, OC shell) to the GRPP
    for (int ioc = 0; ioc < n_oc_shells; ioc++) {
        libgrpp_grpp_add_outercore_potential(grpp, buf_oc_potentials[ioc], buf_oc_shells[ioc]);
    }

    fclose(inp_file);

    return grpp;
}


void print_grpp(FILE *out, libgrpp_grpp_t *grpp)
{
    char ang_mom_labels[] = "SPDFGHIKLMNOPQ";

    printf("\toutercore pseudospinors:\n\n");

    for (int ioc = 0; ioc < grpp->n_oc_shells; ioc++) {
        libgrpp_shell_t *shell = grpp->oc_shells[ioc];
        char L_symbol = ang_mom_labels[shell->L];

        for (int i = 0; i < shell->num_primitives; i++) {
            fprintf(out, "  %c%18.8e%20.8e\n", (i == 0) ? L_symbol : ' ', shell->alpha[i], shell->coeffs[i]);
        }

        printf("\n");
    }

    printf("\t\tradially-local potential:\n\n");
    print_rpp_expansion(stdout, grpp->U_L);
    printf("\n");

    printf("\t\tsemi-local averaged potentials:\n\n");
    for (int L = 0; L < grpp->n_arep; L++) {
        print_rpp_expansion(stdout, grpp->U_arep[L]);
        printf("\n");
    }

    printf("\t\tsemi-local spin-orbit potentials:\n\n");
    for (int L = 1; L < grpp->n_esop; L++) {
        print_rpp_expansion(stdout, grpp->U_esop[L]);
        printf("\n");
    }

    printf("\t\tnon-local outercore potentials:\n\n");
    for (int ioc = 0; ioc < grpp->n_oc_shells; ioc++) {
        print_rpp_expansion(stdout, grpp->U_oc[ioc]);
        printf("\n");
    }
}


void print_rpp_expansion(FILE *out, libgrpp_potential_t *ecp)
{
    char ang_mom_labels[] = "SPDFGHIKLMNOPQ";
    char L_symbol = ang_mom_labels[ecp->L];
    char J_symbol[8];

    if (ecp->J > 0) {
        sprintf(J_symbol, "%2d/2", ecp->J);
    }
    else {
        sprintf(J_symbol, "    ");
    }

    for (int i = 0; i < ecp->num_primitives; i++) {
        fprintf(out, "  %c%4s%4d%20.8e%20.8e\n",
                (i == 0) ? L_symbol : ' ',
                (i == 0) ? J_symbol : "",
                ecp->powers[i], ecp->alpha[i], ecp->coeffs[i]);
    }
}


int skip_line(FILE *fp)
{
    int c;

    while (c = fgetc(fp), c != '\n' && c != EOF);

    return c;
}
